File: backend/app/schemas/apify_key.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class ApifyKeyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    label: str
    token: str
    is_default: bool
    remark: Optional[str] = None
    exhausted_at: Optional[datetime] = None
    email_account_id: Optional[int] = None
    email_account_email: Optional[str] = None
    email_account_verification_email: Optional[str] = None
    apify_full_name: Optional[str] = None
    apify_username: Optional[str] = None
    apify_user_id: Optional[str] = None
    apify_registered_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ApifyKeyCreate(BaseModel):
    label: str = Field(..., min_length=1, max_length=200)
    token: str = Field(..., min_length=1, max_length=500)
    is_default: bool = False
    remark: Optional[str] = Field(None, max_length=500)
    email_account_id: Optional[int] = None
    apify_full_name: Optional[str] = Field(None, max_length=128)
    apify_username: Optional[str] = Field(None, max_length=128)
    apify_user_id: Optional[str] = Field(None, max_length=128)
    apify_registered_at: Optional[datetime] = None


class ApifyKeyUpdate(BaseModel):
    label: Optional[str] = Field(None, min_length=1, max_length=200)
    token: Optional[str] = Field(None, min_length=1, max_length=500)
    remark: Optional[str] = Field(None, max_length=500)
    email_account_id: Optional[int] = None
    apify_full_name: Optional[str] = Field(None, max_length=128)
    apify_username: Optional[str] = Field(None, max_length=128)
    apify_user_id: Optional[str] = Field(None, max_length=128)
    apify_registered_at: Optional[datetime] = None

File: frontend/src/api/apifyKey.ts
import http from './http'

export interface ApifyKey {
  id: number
  label: string
  token: string
  is_default: boolean
  remark?: string | null
  exhausted_at?: string | null
  email_account_id?: number | null
  email_account_email?: string | null
  email_account_verification_email?: string | null
  created_at: string
  updated_at: string
}

export interface ApifyKeyCreate {
  label: string
  token: string
  is_default?: boolean
  remark?: string | null
  email_account_id?: number | null
}

export interface ApifyKeyUpdate {
  label?: string
  token?: string
  remark?: string | null
  email_account_id?: number | null
}

export const apifyKeyApi = {
  list: () => http.get<unknown, ApifyKey[]>('/scraper/apify-keys'),
  create: (data: ApifyKeyCreate) => http.post<unknown, ApifyKey>('/scraper/apify-keys', data),
  update: (id: number, data: ApifyKeyUpdate) =>
    http.put<unknown, ApifyKey>(`/scraper/apify-keys/${id}`, data),
  remove: (id: number) => http.delete<unknown, { ok: boolean }>(`/scraper/apify-keys/${id}`),
  setDefault: (id: number) =>
    http.post<unknown, ApifyKey>(`/scraper/apify-keys/${id}/set-default`, {}),
  markExhausted: (id: number) =>
    http.post<unknown, ApifyKey>(`/scraper/apify-keys/${id}/mark-exhausted`, {}),
  unmarkExhausted: (id: number) =>
    http.post<unknown, ApifyKey>(`/scraper/apify-keys/${id}/unmark-exhausted`, {})
}
